tests/runtimes/test_runtime_handler.py

```diff
@@ -0,0 +1,690 @@
+    )
+
+
+def test_list_daskjob_resources(k8s_helper_mock):
+    pods = _mock_list_daskjob_pods(k8s_helper_mock)
+    services = _create_daskjob_service_mocks(k8s_helper_mock)
+    _assert_runtime_handler_list_resources(
+        RuntimeKinds.dask,
+        k8s_helper_mock,
+        expected_pods=pods,
+        expected_services=services,
+    )
+
+
+def test_list_mpijob_resources(k8s_helper_mock):
+    config.mpijob_crd_version = MPIJobCRDVersions.v1
+    crds = _mock_list_mpijob_crds(k8s_helper_mock)
+
+    # there's currently a bug (fix was merged but not released https://github.com/kubeflow/mpi-operator/pull/271)
+    # that causes mpijob's pods to not being labels with the given (MLRun's) labels - this prevents list resources from
+    # finding the pods, so we're simulating the same thing here
+    k8s_helper_mock.list_pods.return_value = []
+    _assert_runtime_handler_list_resources(
+        RuntimeKinds.mpijob, k8s_helper_mock, expected_crds=crds
+    )
+
+
+def test_list_sparkjob_resources(k8s_helper_mock):
+    crds = _mock_list_sparkjob_crds(k8s_helper_mock)
+    pods = _mock_list_daskjob_pods(k8s_helper_mock)
+    _assert_runtime_handler_list_resources(
+        RuntimeKinds.spark, k8s_helper_mock, expected_crds=crds, expected_pods=pods
+    )
+
+
+def _assert_runtime_handler_list_resources(
+    runtime_kind,
+    k8s_helper_mock,
+    expected_crds=None,
+    expected_pods=None,
+    expected_services=None,
+):
+    runtime_handler = get_runtime_handler(runtime_kind)
+    resources = runtime_handler.list_resources()
+    crd_group, crd_version, crd_plural = runtime_handler._get_crd_info()
+    k8s_helper_mock.list_pods.assert_called_once_with(
+        k8s_helper_mock.resolve_namespace(),
+        selector=runtime_handler._get_default_label_selector(),
+    )
+    if expected_crds:
+        k8s_helper_mock.crdapi.list_namespaced_custom_object.assert_called_once_with(
+            crd_group,
+            crd_version,
+            k8s_helper_mock.resolve_namespace(),
+            crd_plural,
+            label_selector=runtime_handler._get_default_label_selector(),
+        )
+    if expected_services:
+        k8s_helper_mock.v1api.list_namespaced_service.assert_called_once_with(
+            k8s_helper_mock.resolve_namespace(),
+            label_selector=runtime_handler._get_default_label_selector(),
+        )
+    _assert_list_resources_response(
+        resources,
+        expected_crds=expected_crds,
+        expected_pods=expected_pods,
+        expected_services=expected_services,
+    )
+
+
+def _assert_list_resources_response(
+    resources, expected_crds=None, expected_pods=None, expected_services=None
+):
+    if expected_crds is None:
+        expected_crds = []
+    else:
+        expected_crds = expected_crds["items"]
+    if expected_pods is None:
+        expected_pods = []
+    if expected_services is None:
+        expected_services = []
+    else:
+        expected_services = expected_services.items
```

Comment:
Is there a reason in this whole block we don't always get a list? (sometimes a dict and sometimes an object)
If they were all lists, all these lines could be made simpler and look nicer as:
`expected_pods = expected_pods or []`

Reply:
Basically the reason is that also in the k8s client the API is different, e.g. `list_namespaced_custom_object` returns a dict with an `items` key which holds the list of custom objects, while `list_namespaced_service` returns an object with an `items` attribute which holds the list of services...
Practically you're right that in this point I don't care about the mocked objects, but just about the list of expected objects, so I changed it so now it looks nice as you suggested